### PR TITLE
feat(p2p/metrics): add `MessageSendDelay` metrics [WIP]

### PR DESCRIPTION
**Pull request by cason**
This is an EXPERIMENT in order to assess some undesired behavior of the p2p communication layer. The idea is to run testnets/e2e tests with this additional metric in order to retrieve some result.

The solution is not optimized and SHOULD NOT BE MERGED at the moment.

---

This metric computes the average send delay for messages, per destination peer (ID) and channel (ID).

The metric is computed properly for the `peer.Send()` method because it is blocking. Since `peer.TrySend()` method is not blocking, we artificially assume a 10s duration when if fails (because the send queue is full).

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

**Comment by melekes** [comment]
What triggered this change? Is there an issue?

File: p2p/metrics.go [hunk 1]
```diff
@@ -34,11 +35,15 @@ type Metrics struct {
...
 	MessageSendBytesTotal metrics.Counter `metrics_labels:"message_type"`
+	// Average delay for sending messages to a peer in a channel.
+	MessageSendDelaySeconds metrics.Gauge `metrics_labels:"peer_id, channel_id"`
 }
 
 type peerPendingMetricsCache struct {
 	mtx             sync.Mutex
 	perMessageCache map[reflect.Type]*peerPendingMetricsCacheEntry
+
+	perChannelChache map[byte]*peerPendingDelayMetrics
```
Comment (melekes) [inline p2p/metrics.go:46]:
```suggestion
	perChannelCache map[byte]*peerPendingDelayMetrics
```

**Comment by cason** [comment]
> What triggered this change? Is there an issue?

Sorry for not explaining. This is a experiment that me and @sergio-mena are running.

File: p2p/metrics.go [hunk 2]
```diff
@@ -79,6 +94,21 @@ func (c *peerPendingMetricsCache) AddPendingRecvBytes(msgType reflect.Type, addB
...
 
+func (c *peerPendingMetricsCache) AddPendingMessageSendDelay(chID byte, delay time.Duration) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	if entry, ok := c.perChannelChache[chID]; ok {
+		entry.count++
+		entry.sendDelay += delay
+	} else {
+		c.perChannelChache[chID] = &peerPendingDelayMetrics{
+			chID:      chID,
```
Comment (sergio-mena) [inline p2p/metrics.go:105]: Why do we need this field here (we have it as the map's key)?

File: p2p/peer.go [hunk 3]
```diff
@@ -256,15 +256,39 @@ func (p *peer) Status() cmtconn.ConnectionStatus {
 //
 // thread safe.
 func (p *peer) Send(e Envelope) bool {
-	return p.send(e.ChannelID, e.Message, p.mconn.Send)
+	start := time.Now()
```
Comment (sergio-mena) [inline p2p/peer.go:259]: `cmttime`? I know it doesn't make any difference, as it's not used as wallclock time, but to calculate duration, but with `cmttime` the code is more homogeneous.

File: p2p/metrics.go (hunk 2 above)
Comment (cason) [reply p2p/metrics.go:105]: True, ehe.

File: p2p/peer.go (hunk 3 above)
Comment (cason) [reply p2p/peer.go:259]: We don't use `cmttime` in this file, but this is a good point. We lose the non-monotonic part, but this shouldn't be a problem.

File: p2p/peer.go [hunk 4]
```diff
@@ -388,6 +405,12 @@ func (p *peer) metricsReporter() {
 						entry.pendingRecvBytes = 0
 					}
 				}
+				for chID, entry := range p.pendingMetrics.perChannelCache {
+					p.metrics.MessageAverageSendDelay.
+						With("peer_id", string(p.ID())).
+						With("channel_id", string(chID)).
+						Set(entry.sendDelay.Seconds() / float64(entry.count))
+				}
```
Comment (sergio-mena) [inline p2p/peer.go:413]:
```suggestion
				}
				p.pendingMetrics.perChannelCache = make(map[byte]*peerPendingDelayMetrics)
```
As we want a gauge, we need to reset every time we report

**Comment by sergio-mena** [comment]
I think we can close this PR, @cason ?

**Pull request closed without merging**